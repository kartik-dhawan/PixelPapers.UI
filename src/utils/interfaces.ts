export interface ChildrenProps {
  children: React.ReactNode
}

export type ContentfulFields = {
  [key: string]: any
}

export interface ThumbnailItem {
  fields: {
    blogMetaTitle: string
    blogTravelLocation: string
    blogType: string
    blogMetaImage: any
  }
  sys: any
}
export interface SocialMediaLink {
  id: number
  label: string
  url: string
  icon: string
}

export interface TravelBlogPost {
  blogTitle?: string
  blogMetaImage?: any
  blogMetaTitle?: string
  blogTravelLocation?: string
  blogType?: string
  blogUrlSlug?: string
  blogInitialDescription?: any
  blogThemeColor?: string
}

export interface ContentfulTravelBlog {
  fields: TravelBlogPost
  sys: any
  metadata: any
}
