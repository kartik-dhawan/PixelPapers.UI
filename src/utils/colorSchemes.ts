export const GLOBAL_COLORS = {
  TEXT_PRIMARY_DARK: "#222",
  TEXT_PRIMARY_LIGHTER: "#666",
  TEXT_PRIMARY_LIGHT: "#d9d9d9",
  BACKGROUND_PRIMARY_LIGHT: "#fefefe",
  BACKGROUND_PRIMARY_DARK: "#111",
  COMMON_FOOTER_BACKGRUND_DARK: "#383b3e",
  BLOG_SUB_COMPONENT_BACKGROUND_LIGHTER: "#efefef",
}

export const PAGE_LEVEL_COLORS = {
  PAGE_404: {
    BACKGROUND_PRIMARY_LIGHT: "#226bb152",
    TEXT_PRIMARY_LIGHT: "#444",
    TEXT_PRIMARY_DARK: "#1c354c",
  },
  TRAVEL: {
    BACKGROUND_PRIMARY_LIGHT: "#fdeded",
    BACKGROUND_PRIMARY_DARK: "#864040",
    TEXT_PRIMARY_LIGHTER: "#444",
  },
}
