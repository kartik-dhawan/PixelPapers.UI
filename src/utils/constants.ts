export const CURATED_BY_LINK_DEFAULT_TEXT = "kartik-dhawan"

export const AVERAGE_TIME_TO_READ_A_BLOG /**In minutes */ = 7
