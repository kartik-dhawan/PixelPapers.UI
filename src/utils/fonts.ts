import { DM_Sans, Anton, Libre_Barcode_39 } from "next/font/google"

export const dmSans = DM_Sans({
  weight: ["200", "300", "400", "500", "600", "700"],
  subsets: ["latin"],
  display: "swap",
  preload: true,
})

export const barcode = Libre_Barcode_39({
  weight: ["400"],
  subsets: ["latin"],
  display: "swap",
})

export const anton = Anton({
  weight: ["400"],
  subsets: ["latin"],
  display: "swap",
})
