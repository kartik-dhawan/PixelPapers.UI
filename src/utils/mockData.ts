export const TRAVEL_PAGE_THUMBNAIL_CARD_DATA = {
  blogTitle: "Top 10 Things to do on your next trip to Jibhi/Jalori Pass.",
  blogMetaImage: {
    metadata: {
      tags: [],
    },
    sys: {
      space: {
        sys: {
          type: "Link",
          linkType: "Space",
          id: "1v98waf66sfj",
        },
      },
      id: "6TVtXnjJIZ812Bxm6VBqz5",
      type: "Asset",
      createdAt: "2023-11-21T17:39:08.757Z",
      updatedAt: "2023-11-21T17:39:17.165Z",
      environment: {
        sys: {
          id: "master",
          type: "Link",
          linkType: "Environment",
        },
      },
      revision: 2,
      locale: "en-US",
    },
    fields: {
      title: "Jibhi_Jalori",
      description:
        "A picture of sunset from Raghupur fort close to jalori pass",
      file: {
        url: "//images.ctfassets.net/1v98waf66sfj/6TVtXnjJIZ812Bxm6VBqz5/5b89ce3dced278570a2a663453d69a65/IMG_3749.webp",
        details: {
          size: 71536,
          image: {
            width: 3024,
            height: 4032,
          },
        },
        fileName: "IMG_3749.webp",
        contentType: "image/webp",
      },
    },
  },
  blogMetaTitle: "Jibhi - Jalori",
  blogTravelLocation: "Himachal Pradesh",
  blogType: "Travel Itinerary",
}
