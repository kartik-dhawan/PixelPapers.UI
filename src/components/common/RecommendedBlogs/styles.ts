import { GLOBAL_COLORS } from "@/utils/colorSchemes"

export const styles = {
  recommendationsSectionWrapper: {
    padding: "2rem",
    boxSizing: "border-box",
    display: "flex",
    justifyContent: "center",
    width: "100vw",
    backgroundColor: GLOBAL_COLORS.RECOMMENDATIONS_BACKGROUND_LIGHTER,
  },
  recommendationsCard: {
    margin: "1rem 0rem",
    "& a": {
      textDecoration: "none",
      color: GLOBAL_COLORS.TEXT_PRIMARY_DARK,
    },
  },
  recommendationsCardImageWrapper: {
    position: "relative",
    maxWidth: "100%",
    height: {
      xs: "200px",
      sm: "250px",
      md: "200px",
      xl: "250px",
    },
    width: "100%",
  },
  recommendationsCardImage: {
    objectFit: "cover",
    objectPosition: {
      xs: "0% 60%",
      sm: "0% 70%",
    },
    boxSizing: "border-box",
  },
  recommendationsCardTextContent: {
    display: "flex",
    flexDirection: "column",
    gap: "8px",
    padding: {
      xs: "1rem 0.5rem",
      sm: "1rem 0rem 1rem 2rem",
      md: "1rem 0.5rem",
    },
    boxSizing: "border-box",
  },
  recommendationsCardAuthorUsername: {
    fontSize: {
      xs: "14px",
      sm: "16px",
      md: "14px",
      xl: "18px",
    },
  },
  recommendationsCardBlogTitle: {
    margin: "4px 0px",
    fontSize: {
      xs: "22px",
      sm: "26px",
      md: "22px",
      xl: "26px",
    },
    textOverflow: "ellipsis",
    whiteSpace: "nowrap",
    overflow: "hidden",
    letterSpacing: "-1px",
    fontWeight: 600,
  },
  recommendationsCardBlogDescription: {
    fontWeight: 300,
    display: "-webkit-box",
    overflow: "hidden",
    WebkitBoxOrient: "vertical",
    fontSize: {
      xs: "18px",
      sm: "20px",
      md: "18px",
      xl: "22px",
    },
    color: GLOBAL_COLORS.TEXT_PRIMARY_LIGHTER,
  },
  recommendationsCardBlogMetaInfo: {
    fontSize: {
      xs: "14px",
      sm: "16px",
      md: "14px",
      xl: "18px",
    },
    color: GLOBAL_COLORS.TEXT_PRIMARY_LIGHTER,
  },
}
