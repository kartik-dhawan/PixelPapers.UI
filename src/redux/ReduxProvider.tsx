"use client"

/**
 *
 * This component acts as a parent for client side components that require redux store
 * Wrap the component with this 'Redux Provider' to use state management on client side
 *
 * the updates made on the client side in the store will not be reflected in the server side components
 *
 * SERVER SIDE:
 * - Dispatching: store.dispatch(functionName(payload))
 * - Getting data from store: store.getState().sliaceName.stateName
 *
 * CLIENT SIDE:
 * - Dispatching: dispatch(functionName(payload)) [useDispatch()]
 * - Getting data from store: useSelector((state)=>state.slice.statename)
 *
 * Ref: https://www.youtube.com/watch?v=dRLjoT4r-jc
 */

import { Provider } from "react-redux"
import { store } from "./store"
import { ChildrenProps } from "@/utils/interfaces"

const ReduxProvider = ({ children }: ChildrenProps) => {
  return <Provider store={store}>{children}</Provider>
}

export default ReduxProvider
