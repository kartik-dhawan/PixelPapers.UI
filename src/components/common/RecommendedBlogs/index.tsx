"use client"

import { RootType } from "@/redux/store"
import {
  ANONYMOUS_AUTHOR_DATA,
  AVERAGE_TIME_TO_READ_A_BLOG,
} from "@/utils/constants"
import { dmSans } from "@/utils/fonts"
import { ChildrenType, TravelBlogContentfulObject } from "@/utils/interfaces"
import { secondsToMinutes } from "@/utils/methods"
import {
  Avatar,
  Box,
  Divider,
  Grid,
  Link,
  Skeleton,
  Stack,
} from "@mui/material"
import Image from "next/image"
import { useSelector } from "react-redux"
import { styles } from "./styles"
import { useState } from "react"

interface RecommendedBlogsProps {
  blogs: TravelBlogContentfulObject[]
  fullWidth?: boolean
  path?: `/${string}`
  children?: ChildrenType
}

/**
 * A custom reusable component for recommendation section.
 * @param {RecommendedBlogsProps} Props
 * ** blogs - the blogs data
 * ** fullWidth - should the blog cards be full width or cards in multiple columns
 * ** path - the path to append to the blog url like '/travel' will be appended to the sluh url /[ghosts-of-shoja]
 * ** children - Children  of this component gets rendered as the title
 *
 */

const RecommendedBlogs = ({
  blogs,
  fullWidth = false, // default value false, it will render square cards
  path,
  children,
}: RecommendedBlogsProps) => {
  const rid = "recommendations"

  const { content } = useSelector((state: RootType) => state.contentSlice)

  const [imageLoadingStatus, setImageLoadingStatus] = useState(false)

  const imageLoadingHandler = () => {
    setImageLoadingStatus(true)
  }

  return (
    <Box
      className={rid + "SectionWrapper " + dmSans.className}
      sx={styles.recommendationsSectionWrapper}
    >
      <>{children}</>
      <Grid
        container
        spacing={{ xs: 2, sm: 4, md: 4 }}
        className={rid + "CardContainer"}
        sx={{
          maxWidth: {
            xs: "720px",
            md: fullWidth ? "880px" : "720px",
            xl: "900px",
          },
        }}
      >
        {blogs.map((item: TravelBlogContentfulObject) => {
          return (
            <Grid
              item
              className={rid + "Card"}
              key={item.sys.id}
              xs={12}
              md={fullWidth ? 12 : 6}
              sx={styles.recommendationsCard}
            >
              <Link
                href={
                  path
                    ? `${path + "/" + item.fields.blogUrlSlug}`
                    : item.fields.blogUrlSlug ?? "#!"
                }
                rel="noreferrer"
                aria-label={`A link to open the blog titled: ${item.fields.blogMetaTitle}, in next tab`}
              >
                <Grid
                  className={rid + "CardContentWrapper"}
                  container
                  sx={{ height: "100%" }}
                >
                  <Grid
                    item
                    xs={12}
                    sm={5}
                    md={fullWidth ? 5 : 12}
                    className={rid + "CardImageWrapper"}
                    sx={{
                      ...styles.recommendationsCardImageWrapper,
                      "& img": {
                        ...styles.recommendationsCardImage,
                        padding: fullWidth
                          ? {
                              sm: "8px",
                              md: "1rem",
                            }
                          : {},
                      },
                    }}
                  >
                    {!imageLoadingStatus && (
                      <Box
                        padding="1rem"
                        boxSizing="border-box"
                        height="100%"
                        width="100%"
                      >
                        <Skeleton
                          variant="rectangular"
                          height="100%"
                          width="100%"
                        />
                      </Box>
                    )}
                    <Image
                      className={rid + "CardImage"}
                      src={`https:${item.fields.blogMetaImage.fields.file.url}`}
                      alt=""
                      fill
                      onLoad={imageLoadingHandler}
                    />
                  </Grid>
                  <Grid
                    item
                    xs={12}
                    sm={7}
                    md={fullWidth ? 7 : 12}
                    className={rid + "CardTextContent"}
                    sx={styles.recommendationsCardTextContent}
                  >
                    <Stack
                      direction="row"
                      alignItems="center"
                      gap="8px"
                      margin="0px 0px"
                      className={rid + "CardAuthorDetails"}
                    >
                      <Avatar
                        src={`https:${
                          item.fields.blogAuthorData
                            ? item.fields.authorImage?.fields.file.url ?? ""
                            : content.myProfileImage?.fields.file.url
                        }`}
                        alt={"Author of the blog"}
                        sx={{
                          bgcolor: item.fields.blogThemeColor,
                          height: "24px",
                          width: "24px",
                        }}
                      />
                      <Box sx={styles.recommendationsCardAuthorUsername}>
                        {item.fields.blogAuthorData
                          ? item.fields.blogAuthorData?.username ??
                            ANONYMOUS_AUTHOR_DATA.USERNAME
                          : content.myAuthorProfileData?.username ??
                            ANONYMOUS_AUTHOR_DATA.USERNAME}
                      </Box>
                    </Stack>
                    <Box className={rid + "CardBlogHead"}>
                      <Box
                        component="h3"
                        className={rid + "CardBlogTitle"}
                        sx={styles.recommendationsCardBlogTitle}
                      >
                        {item.fields.blogMetaTitle}
                      </Box>
                      <Box
                        className={rid + "CardBlogDescription"}
                        sx={{
                          ...styles.recommendationsCardBlogDescription,
                          WebkitLineClamp: {
                            xs: fullWidth ? 3 : 2,
                            sm: 3,
                            md: fullWidth ? 3 : 2,
                          },
                        }}
                      >
                        {item.fields.initialBlogDescription}
                      </Box>
                    </Box>
                    <Stack
                      direction="row"
                      justifyContent="space-between"
                      className={rid + "CardBlogMetaInformation"}
                      sx={{
                        ...styles.recommendationsCardBlogMetaInfo,
                        marginTop: {
                          xs: fullWidth ? "auto" : "1rem",
                          sm: "auto",
                          md: fullWidth ? "auto" : "1rem",
                        },
                      }}
                    >
                      <span>
                        {secondsToMinutes(
                          item.fields.blogMetaInformation
                            ?.timeToReadInSeconds ??
                            AVERAGE_TIME_TO_READ_A_BLOG * 60,
                        )}
                        &nbsp;min read
                      </span>
                      <span style={{ textTransform: "uppercase" }}>
                        {item.fields.blogType}
                      </span>
                    </Stack>
                  </Grid>
                </Grid>
              </Link>
              <Divider
                className={rid + "CardDivider"}
                sx={{
                  marginTop: "16px",
                  display: { xs: "flex", md: fullWidth ? "flex" : "none" },
                }}
              />
            </Grid>
          )
        })}
      </Grid>
    </Box>
  )
}

export default RecommendedBlogs
