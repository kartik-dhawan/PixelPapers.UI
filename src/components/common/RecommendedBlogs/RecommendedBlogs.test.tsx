import "@testing-library/jest-dom"
import renderer from "react-test-renderer"
import { BLOG_DATA_MOCK } from "@/utils/mockData"
import RecommendedBlogs from "."
import ReduxProvider from "@/redux/ReduxProvider"

describe("Custom RecommendedBlogs component", () => {
  it("-- should match the snapshot", () => {
    const tree = renderer.create(
      <ReduxProvider>
        <RecommendedBlogs blogs={[BLOG_DATA_MOCK]} path="/" fullWidth />,
      </ReduxProvider>,
    )

    expect(tree).toMatchSnapshot()
  })
})
