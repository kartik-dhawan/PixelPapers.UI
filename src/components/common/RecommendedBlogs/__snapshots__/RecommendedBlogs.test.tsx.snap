// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom RecommendedBlogs component -- should match the snapshot 1`] = `
[
  <div
    className="recommendationsSectionWrapper className MuiBox-root css-15un6cr"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-spacing-sm-4 MuiGrid-spacing-md-4 recommendationsCardContainer css-it53zt-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-12 recommendationsCard css-1xmv5g-MuiGrid-root"
      >
        <a
          aria-label="A link to open the blog titled: The Labyrith of Jodhpur, in next tab"
          className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1xylxj1-MuiTypography-root-MuiLink-root"
          href="//labyrinth-of-jodhpur"
          onBlur={[Function]}
          onFocus={[Function]}
          rel="noreferrer"
        >
          <div
            className="MuiGrid-root MuiGrid-container recommendationsCardContentWrapper css-1sgmr83-MuiGrid-root"
          >
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5 MuiGrid-grid-md-5 recommendationsCardImageWrapper css-4ez996-MuiGrid-root"
            >
              <div
                className="MuiBox-root css-kq8hvf"
              >
                <span
                  className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-153n0d3-MuiSkeleton-root"
                  style={
                    {
                      "height": "100%",
                      "width": "100%",
                    }
                  }
                />
              </div>
              <img
                alt=""
                className="recommendationsCardImage"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                onError={[Function]}
                onLoad={[Function]}
                sizes="100vw"
                src="/_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=3840&q=75"
                srcSet="/_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimages%2FSnapseed.webp&w=3840&q=75 3840w"
                style={
                  {
                    "bottom": 0,
                    "color": "transparent",
                    "height": "100%",
                    "left": 0,
                    "objectFit": undefined,
                    "objectPosition": undefined,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "width": "100%",
                  }
                }
              />
            </div>
            <div
              className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-7 MuiGrid-grid-md-7 recommendationsCardTextContent css-1gz10iw-MuiGrid-root"
            >
              <div
                className="MuiStack-root recommendationsCardAuthorDetails css-1s4zrlh-MuiStack-root"
              >
                <div
                  className="MuiAvatar-root MuiAvatar-circular css-u9up2x-MuiAvatar-root"
                >
                  <img
                    alt="Author of the blog"
                    className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
                    src="https:undefined"
                  />
                </div>
                <div
                  className="MuiBox-root css-1pvqewc"
                >
                  Anonymous
                </div>
              </div>
              <div
                className="recommendationsCardBlogHead MuiBox-root css-0"
              >
                <h3
                  className="recommendationsCardBlogTitle MuiBox-root css-1rsd7ub"
                >
                  The Labyrith of Jodhpur
                </h3>
                <div
                  className="recommendationsCardBlogDescription MuiBox-root css-13tgwgo"
                >
                  In the heart of Rajasthan, Jodhpur stands as a vibrant testament to its regal past. The majestic Mehrangarh Fort and the azure hues of the Blue City vividly recall the cultural splendor that I had the privilege to explore firsthand.
                </div>
              </div>
              <div
                className="MuiStack-root recommendationsCardBlogMetaInformation css-1rg0q2c-MuiStack-root"
              >
                <span>
                  7
                  Â min read
                </span>
                <span
                  style={
                    {
                      "textTransform": "uppercase",
                    }
                  }
                >
                  article
                </span>
              </div>
            </div>
          </div>
        </a>
        <hr
          className="MuiDivider-root MuiDivider-fullWidth recommendationsCardDivider css-ejq953-MuiDivider-root"
        />
      </div>
    </div>
  </div>,
  ",",
]
`;
